src/main/java/dev/blaauwendraad/masker/json/MaskingState.java

```diff
@@ -21,18 +19,19 @@ final class MaskingState implements ValueMaskerContext {
     private int replacementOperationsTotalDifference = 0;
 
     /**
-     * Current JSONPath is represented by a dequeue of segment references.
+     * Current JSONPath is represented by a stack of segment references.
+     * A stack is implemented with an array of the trie nodes that reference the end of the segment
      */
-    private final Deque<JsonPathNode> currentJsonPath;
+    private KeyMatcher.TrieNode[] currentJsonPath = null;
+    private int currentJsonPathIndex = -1;
+    private int currentJsonPathCapacity = 100;
```

Comment:
This can be `currentJsonPath.length`, no need to track it separately

Reply:
Nice one, fixed.